autopush/tests/test_integration.py

```diff
@@ -0,0 +1,629 @@
+        if self.channels:
+            chans = self.channels.keys()
+        else:
+            chans = []
+        hello_dict = dict(messageType="hello", uaid=self.uaid or "",
+                          channelIDs=chans)
+        if self.use_webpush:
+            hello_dict["use_webpush"] = True
+        msg = json.dumps(hello_dict)
+        log.debug("Send: %s", msg)
+        self.ws.send(msg)
+        result = json.loads(self.ws.recv())
+        log.debug("Recv: %s", result)
+        if self.uaid and self.uaid != result["uaid"]:  # pragma: nocover
+            log.debug("Mismatch on re-using uaid. Old: %s, New: %s",
+                      self.uaid, result["uaid"])
+            self.channels = {}
+        self.uaid = result["uaid"]
+        eq_(result["status"], 200)
+        return result
+
+    def register(self, chid=None):
+        chid = chid or str(uuid.uuid4())
+        msg = json.dumps(dict(messageType="register", channelID=chid))
+        log.debug("Send: %s", msg)
+        self.ws.send(msg)
+        result = json.loads(self.ws.recv())
+        log.debug("Recv: %s", result)
+        eq_(result["status"], 200)
+        eq_(result["channelID"], chid)
+        self.channels[chid] = result["pushEndpoint"]
+        return result
+
+    def unregister(self, chid):
+        msg = json.dumps(dict(messageType="unregister", channelID=chid))
+        log.debug("Send: %s", msg)
+        self.ws.send(msg)
+        result = json.loads(self.ws.recv())
+        log.debug("Recv: %s", result)
+        return result
+
+    def send_notification(self, channel=None, version=None, data=None,
+                          use_header=True, status=200, ttl=200):
+        if not channel:
+            channel = random.choice(self.channels.keys())
+
+        endpoint = self.channels[channel]
+        url = urlparse.urlparse(endpoint)
+        http = None
+        if url.scheme == "https":  # pragma: nocover
+            http = httplib.HTTPSConnection(url.netloc)
+        else:
+            http = httplib.HTTPConnection(url.netloc)
+
+        if self.use_webpush:
+            headers = {
+                "Content-Type": "application/octet-stream",
+                "Content-Encoding": "aesgcm-128",
+                "Encryption": self._crypto_key,
+                "Encryption-Key": 'keyid="a1"; key="JcqK-OLkJZlJ3sJJWstJCA"',
+                "TTL": str(ttl),
+            }
+            body = data or ""
+            method = "POST"
+            status = 201
+        else:
+            if data:
+                body = "version=%s&data=%s" % (version or "", data)
+            else:
+                body = "version=%s" % (version or "")
+            if use_header:
+                headers = {"Content-Type": "application/x-www-form-urlencoded"}
+            else:
+                headers = {}
+            method = "PUT"
+
+        log.debug("%s body: %s", method, body)
+        http.request(method, url.path, body, headers)
+        resp = http.getresponse()
```

Comment:
Not entirely sure, I haven't tried sending multiple at once without checking the response.